### PR TITLE
fix size of tree body data

### PR DESCRIPTION
**Pull request by myk002**
I discovered this from dumping the tree body data to make a visual bitmap of the tree structure. When we interpret each element as 8 bits, there were gaps between adjacent trunk tiles and the trunk is not centered in the tree dimensions. With a 16-bit tree body data element, trunks are adjacent and centered, and match the visual representation of the in-game trees.

Why did the bit size increase? Are there more flags now in the enum? I'm not equipped to divine that knowledge.

The impact of this change is that `Designations::markPlant()` will now mark the correct (southeast) tile of 3x3 trees instead of the (unreachable) center tile. Also, autochop will get the right count of trunk tiles for estimating log yield, which should fix its chronic under-estimation issues.

### CODE REVIEW

File: df.plants.xml [hunk 1]
```diff
@@ -24,7 +24,7 @@
         <pointer name='tree_info' type-name='plant_tree_info'/>
     </struct-type>
 
-    <bitfield-type type-name='plant_tree_tile' base-type='uint8_t'>
+    <bitfield-type type-name='plant_tree_tile' base-type='uint16_t'>
```
Comment (ab9rf) [inline df.plants.xml:27]: based on analysis of the `write_file` method, this is correct for the use of `plant_tree_tile` for the `body` element of `plant_tree_info` but not correct for the `roots` element of `plant_tree_info`, which currently (and incorrectly) shares this type

so, in addition to this change, which is correct, we also need to add a new type (perhaps `plant_root_tile`) of base type `uint8_t` and change the type of `plant_tree_info::roots` to that new type

**Review by ab9rf — CHANGES_REQUESTED** [review]

File: df.plants.xml (hunk 1 above)
Comment (myk002) [reply df.plants.xml:27]: done

Comment (lethosor) [reply df.plants.xml:27]: Did this actually need to be a new type (perhaps because it's used in a nested array)? Another option is to set `base-type` on an individual enum field to reflect where an enum value is stored in a field of a nontypical size. I expect that's what DF is really doing here - I doubt Toady made a second enum with identical values for roots.

Comment (myk002) [reply df.plants.xml:27]: How would that look? something like this?
```diff
         <pointer name='body' is-array='true' comment='dimension body_height'>
-            <pointer type-name='plant_tree_tile' is-array='true' comment='dimension dim_x*dim_y'/>
+            <pointer type-name='plant_tree_tile' is-array='true' base-type='uint16_t' comment='dimension dim_x*dim_y'/>
         </pointer>
```

Comment (lethosor) [reply df.plants.xml:27]: It would be a bit more complicated, since you can't add `enum` attributes directly to a `pointer` element. Something like this should work (untested):
```xml
        <pointer name='body' is-array='true' comment='dimension body_height'>
            <pointer is-array='true' comment='dimension dim_x*dim_y'>
                 <enum type-name='plant_tree_tile' base-type='uint16_t'/>
             </pointer>
        </pointer>
```

Comment (ab9rf) [reply df.plants.xml:27]: > Did this actually need to be a new type (perhaps because it's used in a nested array)? Another option is to set `base-type` on an individual enum field to reflect where an enum value is stored in a field of a nontypical size. I expect that's what DF is really doing here - I doubt Toady made a second enum with identical values for roots.

since one of these is `int8_t` and the other `int16_t` it's exceedingly likely that toady actually has two different enum types for these. from what i've seen toady declares all enums as 32-bit (he does not use C++11's "base type" functionality, which results in the compiler defaulting to `int` which is 32 bits on all relevant platforms) and then casts these 32-bit enums down to 16 bits or 8 bits when actually using them based on his own determination (not the compiler's) as to how many bytes are needed. because the root tile enum is allocated at 8 bits per unit but the tree tile enum is allocated at 16 bits per unit, that indicates to me that these are in fact two different types, even if they do have overlaps in interpretation

in other words, i'm quite certain that there are, in fact, two separate types and quite likely two separate enums underlying those types

Comment (lethosor) [reply df.plants.xml:27]: > then casts these 32-bit enums down to 16 bits or 8 bits when actually using them based on his own determination (not the compiler's)

Yes, that's my understanding as well, but the fact that there are cases _at all_ where he uses differently-sized fields for the same enum is evidence that he's not consistent. These two fields happen to be in the same structure, which is more unusual, but it's possible (for instance) that he changed one field from int16 to int8 at one point to save space and overlooked the other one.

Hard to tell without access to the source. I agree there's not much point in changing it now, but I would prefer to avoid adding new types like this in the future if our existing support for `<enum base-type>` can work instead.
